[ci-visibility] Support mocha parallel mode

juan-fernandez opened the pull request:
### What does this PR do?

Add support for `mocha --parallel`. 

Changes:
* Split the `mocha.js` instrumentation file into three: `main.js` and `worker.js`, which correspond to the main process and the worker processes, and `common.js` for common pieces. Additionally extract some common functionality into `utils.js`
* `worker.js` file will only instrument _tests_. Suites, modules and sessions will be handled by `main.js`.
* `main.js` will mostly include everything that `mocha.js` had, but we've added two extra hooks for `lib/nodejs/parallel-buffered-runner.js` and `workerpool` to support parallel mode. 
* Added a hook for `workerpool` library because we need to intercept interprocess communication between mocha's main process and the workers, so we can report the test data.

### Motivation
Closes #4046

We want to be able to support mocha when it's run in parallel mode: https://mochajs.org/#parallel-tests

### Plugin Checklist
<!-- Fill this section if adding or updating a plugin. Remove otherwise. -->

- [x] Unit tests.

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/common.js:
```diff
@@ -0,0 +1,50 @@
+    return addTest.apply(this, arguments)
+  })
+  return Suite
+})
```
these two hooks are not modified at all

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+
+    return runner
+  })
+  return Mocha
```
this hook is basically the same as it was in `mocha.js`

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+    }
+    return runMocha.apply(this, arguments)
+  })
+  return run
```
this is the same as before

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+    return run.apply(this, arguments)
+  })
+
+  return Runner
```
same as before too

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/common.js:
```diff
@@ -0,0 +1,50 @@
+const testToStartLine = new WeakMap()
+
+// mocha-each support
+addHook({
```
ℹ️ these two hooks are not modified at all

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+// skippable and known tests.
+// It is called but skipped in parallel mode. In parallel mode, we'll use
+// the run-helpers hook to grab the data.
+addHook({
```
ℹ️ this hook was not modified

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+
+// Only used in serial mode (no --parallel flag is passed)
+// This hook is used to generate session, module, suite and test events
+addHook({
```
ℹ️ this hook was barely modified, only to add some minor checks

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+// In serial mode, this hook is used to set the correct async resource to the test.
+// In parallel mode, the same hook is executed in the worker, so this needs to be repeated in
+// mocha/worker.js
+addHook({
```
ℹ️ the hook was not modified

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+
+// Only used in parallel mode (--parallel flag is passed)
+// Used to generate suite events and receive test payloads from workers
+addHook({
```
⚠️ This hook was added to support parallel mode

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+
+// Only used in parallel mode (--parallel flag is passed)
+// Used to start and finish test session and test module
+addHook({
```
⚠️ This hook was added to support parallel mode

juan-fernandez commented on packages/datadog-instrumentations/src/helpers/hooks.js:
```diff
@@ -71,6 +71,7 @@ module.exports = {
   'microgateway-core': () => require('../microgateway-core'),
   mocha: () => require('../mocha'),
   'mocha-each': () => require('../mocha'),
+  workerpool: () => require('../mocha'),
```
Self review: check whether we need to disable this for out of test processes

juan-fernandez commented on integration-tests/ci-visibility/run-workerpool.js:
```diff
@@ -0,0 +1,23 @@
+// eslint-disable-next-line
+const workerpool = require('workerpool')
```
ℹ️ to test that running `workerpool` outside of a test does not crash

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/worker.js:
```diff
@@ -0,0 +1,199 @@
+const { addHook, channel, AsyncResource } = require('../helpers/instrument')
```
ℹ️ file for the mocha worker: it's a smaller version of `mocha/main.js`, with support for tests only

rochdev replied:
Since as you mentioned it's basically the same as `main.js`, any way to reuse most of the code? There seems to be quite a lot of duplication between the two otherwise.

juan-fernandez replied:
there are small differences, but I guess I could extract _some_ mocha event handlers like `this.on('test', testHandler`). I'll give that a try, thanks 👍

juan-fernandez replied:
done

uurien commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+  versions: ['>=6.0.0'],
+  file: 'src/WorkerHandler.js'
+}, (workerHandlerPackage) => {
+  shimmer.wrap(workerHandlerPackage.prototype, 'exec', exec => async function (message, [testSuiteAbsolutePath]) {
```
I think that you can't do this, workerpool is using custom [Promise](https://github.com/josdejong/workerpool/blob/54af056a09c08c5ef26ca6c6d066e2db3d480978/src/WorkerHandler.js#L3)

You are replacing custom Promise by the native Promise

juan-fernandez replied:
good catch. Apparently it's inspired by this https://gist.github.com/RubaXa/8501359, which says "is supported as a native interface", but I'm sure if that means what it seems to mean 😆 . In any case, making removing the `async` should be easy

uurien commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,678 @@
+      })
+    })
+
+    const result = await exec.apply(this, arguments)
```
If the promise is rejected, next lines are not going to be executed, is it something expected?

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,481 @@
+  versions: ['>=5.2.0'],
+  file: 'lib/runner.js'
+}, function (Runner, frameworkVersion) {
+  if (patched.has(Runner)) return Runner
```
ℹ️ the logic on this hook was not modified, but we extracted some of the mocha event handlers to functions in `mocha/utils.js`: `getOnStartHandler` etc.

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,481 @@
+      })
+      throw err
+    }
+  })
```
the rejection is now handled @uurien

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/main.js:
```diff
@@ -0,0 +1,481 @@
+  versions: ['>=6.0.0'],
+  file: 'src/WorkerHandler.js'
+}, (workerHandlerPackage) => {
+  shimmer.wrap(workerHandlerPackage.prototype, 'exec', exec => function (message, [testSuiteAbsolutePath]) {
```
also, I avoided using `async` @uurien

juan-fernandez commented on packages/datadog-instrumentations/src/mocha/utils.js:
```diff
@@ -0,0 +1,305 @@
+const {
```
ℹ️ this file now contains functions used both for parallel and serial mode and for the main and worker processes

juan-fernandez commented on integration-tests/ci-visibility.spec.js:
```diff
@@ -2498,44 +2613,7 @@ testFrameworks.forEach(({
           }).catch(done)
         })
       })
-      it('marks the test session as skipped if every suite is skipped', (done) => {
```
ℹ️ this test was duplicated

juan-fernandez commented:
ℹ️ I missed a `require`. I noticed it with some manual tests. I fixed it in https://github.com/DataDog/dd-trace-js/pull/4314/commits/bb2c5fc3555f8683f6a71e147b91a8cd3c8fb3d9 and added an assertion in the e2e tests.

juan-fernandez commented:
could you have a look at the latest commit @uurien ?
